src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,346 @@
+    ensure!(matched.into_iter().all(|ok| ok), ChecksumMismatchedSnafu);
+
+    if len == 1 {
+        // A sequence with a single record must contain only a Full record.
+        if sequence[0] != i32::from(RecordType::Full) {
+            return IllegalSequenceSnafu {
+                error: "Missing Full record",
+            }
+            .fail();
+        }
+    } else {
+        // A sequence with multiple records must start with a First record and end with a Last record.
+        // In addition, the Middle records must in order.
+        let prefix = (0..(len - 1) as i32).collect::<Vec<_>>();
+        ensure!(
+            sequence[0..len - 1] == prefix
+                && sequence.last().unwrap() == &i32::from(RecordType::Last),
+            IllegalSequenceSnafu {
+                error: "Out of order",
+            }
+        );
+    }
+
+    Ok(())
+}
+
+fn record_type(seq: usize, num_records: usize) -> RecordType {
+    if seq == 0 {
+        RecordType::First
+    } else if seq == num_records - 1 {
+        RecordType::Last
+    } else {
+        RecordType::Middle(seq)
+    }
+}
+
+fn build_records(entry: EntryImpl, max_record_size: usize) -> Vec<Record> {
+    let payload = entry.data.len();
+    let checksum = crc32fast::hash(&entry.data);
+
+    if payload <= max_record_size {
+        let record = Record {
+            version: VERSION,
+            tp: RecordType::Full,
+            data: entry.data,
+            entry_id: entry.id,
+            ns: entry.ns,
+            checksum,
+        };
+        return vec![record];
+    }
+
+    let num_records = (payload + max_record_size - 1) / max_record_size;
+    let mut records = Vec::with_capacity(num_records);
+    for i in 0..num_records {
+        let end = max_record_size.min(payload - i);
+        let record = Record {
+            version: VERSION,
+            tp: record_type(i, num_records),
+            data: entry.data[i..end].to_vec(),
+            entry_id: entry.id,
+            ns: entry.ns.clone(),
+            checksum,
+        };
+        records.push(record);
+    }
+    records
+}
+
+pub fn maybe_emit_entry(
+    record: Record,
+    ns: &NamespaceImpl,
+    entry_records: &mut HashMap<EntryId, Vec<Record>>,
+) -> Result<Option<EntryImpl>> {
+    let mut entry = None;
+    match record.tp {
+        RecordType::Full => {
+            entry = Some(EntryImpl::try_from(vec![record])?);
+        }
+        RecordType::Last => {
+            // There must have a sequence prefix before a Last record is read.
+            let mut records =
+                entry_records
+                    .remove(&record.entry_id)
+                    .context(IllegalSequenceSnafu {
+                        error: "Missing sequence prefix",
+                    })?;
+            records.push(record);
+            entry = Some(EntryImpl::try_from(records)?);
+        }
+        _ => {
+            entry_records
+                .entry(record.entry_id)
+                .or_default()
```

Comment:
Only on RecordType::First should this `or_default()` be called.

Reply:
Will differentiate between First and Middle types.